[Logs Explorer] Allow to filter data views by type

flash1293 opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/176673
<img width="594" alt="Screenshot 2024-03-20 at 16 33 04" src="https://github.com/elastic/kibana/assets/1508364/c1ac0c53-6d50-4874-8034-ed05984fd26b">

<img width="584" alt="Screenshot 2024-03-20 at 16 33 19" src="https://github.com/elastic/kibana/assets/1508364/f3bb2aff-c746-4243-b75a-6509c8a6713f">

## Technical notes

* I named this new thing a "filter" instead of a "grouping", as grouping sounds too much like collapsing groups, but still showing all the data, while this is about hiding/showing things based on the category
* In the source state machine, the state is held in the `dataViewsFilter` prop, as it's specific to data views (at least for now)
* In the upstream data views state machine, this is kept under the `filter` property
* The existing search logic is extended to not only search by name, but also by filter

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

flash1293 commented:
/ci

elasticmachine commented:
Pinging @elastic/obs-ux-logs-team (Team:obs-ux-logs)

flash1293 commented:
@tonyghiani would you like to see a functional test for this kind of feature? Trying to be mindful of the runtime of the functional test suite here...

tonyghiani commented:
> @tonyghiani would you like to see a functional test for this kind of feature? Trying to be mindful of the runtime of the functional test suite here...

Would be great if you could add a test on the data_source_selector suites for both stateful/serverless that verifies the displayed results after a filter selection! There should be almost everything you need already in the page object apart from an abstraction to update the selector value.

flash1293 commented:
Added a test - let's see whether I broke something else.

A bunch of tests checked for the context menu title that doesn't exist anymore - I removed these checks completely as IMHO the check for the dataviews itself is implicitly making clear what's going on. Let me know whether I miss something here - what we could do instead is to check for the presence of the filter object, but as mentioned, I don't think it adds a lot of value.

tonyghiani commented:
> Added a test - let's see whether I broke something else.
> 
> A bunch of tests checked for the context menu title that doesn't exist anymore - I removed these checks completely as IMHO the check for the dataviews itself is implicitly making clear what's going on. Let me know whether I miss something here - what we could do instead is to check for the presence of the filter object, but as mentioned, I don't think it adds a lot of value.

Thanks, I see there are still some test failures so I guess you are still working on this, I'll start the review as soon as it goes green

tonyghiani commented on x-pack/plugins/observability_solution/logs_explorer/public/state_machines/data_views/src/services/data_views_service.ts:
```diff
@@ -18,22 +24,26 @@ interface LoadDataViewsServiceDeps {
   };
 
-export const searchDataViews = (dataViews: DataViewDescriptor[], search: DataViewsSearchParams) => {
+export const searchDataViews = (
+  dataViews: DataViewDescriptor[],
+  search: DataViewsSearchParams,
+  filter: DataViewsFilterParams
+) => {
   const { name, sortOrder } = search;
+  const { dataType } = filter;
 
   return dataViews
     .filter((dataView) => Boolean(dataView.name?.includes(name ?? '')))
+    .filter((dataView) => !dataType || Boolean(dataView.dataType === dataType))
     .sort(createComparatorByField('name', sortOrder));
 };
```
**suggestion**: this might be a good point to do a small refactor and introduce an object parameter for the search criteria, although is not there yet, we might need to filter only one day without passing the search, and having them listed as inline parameters might me more contrived. 
```ts
interface DataViewsSearchCriteria {
  filter?: DataViewsFilterParams;
  search?: DataViewsSearchParams;
}

export const searchDataViews = (
  dataViews: DataViewDescriptor[],
  searchCriteria: DataViewsSearchCriteria
) => {
  const { filter = {}, search = {} } = searchCriteria;
  const { name, sortOrder = 'asc' } = search;
  const { dataType } = filter;

  return dataViews
    .filter((dataView) => Boolean(dataView.name?.includes(name ?? '')))
    .filter((dataView) => !dataType || Boolean(dataView.dataType === dataType))
    .sort(createComparatorByField('name', sortOrder));
};
```

tonyghiani commented on x-pack/plugins/observability_solution/logs_explorer/public/components/data_source_selector/data_source_selector.tsx:
```diff
@@ -235,7 +241,17 @@ export function DataSourceSelector({
         onSearch={searchByName}
         onSort={sortByOrder}
         isLoading={isSearchingIntegrations || isLoadingUncategorized}
+        filterComponent={
+          tabId === DATA_VIEWS_TAB_ID && (
+            <DataViewsFilter
+              filter={dataViewsFilter}
+              count={dataViewCount}
+              onFilter={filterByType}
+            />
+          )
+        }
```
**note**: in upcoming work we'll need to differ available search/sort/filter controls depending on the context.
This looks good for the scope of this work, in an open PR I might refactor this to be a compound component which allows better composition depending on the tab, but as I said LGTM.

tonyghiani commented on x-pack/plugins/observability_solution/logs_explorer/public/components/data_source_selector/sub_components/data_view_filter.tsx:
```diff
@@ -0,0 +1,109 @@
+                    'data-test-subj': 'logsExplorerDataSourceSelectorDataViewTypeLogs',
+                    icon: filter.dataType === logsDataViewType ? 'check' : 'empty',
+                    name: logsDataViewTypeLabel,
+                    onClick: createSelectTypeFilter('logs'),
```
**nit**: you might want to use the defined constant `logsDataViewType` here

tonyghiani commented on x-pack/plugins/observability_solution/logs_explorer/public/components/data_source_selector/sub_components/data_view_filter.tsx:
```diff
@@ -0,0 +1,109 @@
+                items: [
+                  {
+                    'data-test-subj': 'logsExplorerDataSourceSelectorDataViewTypeAll',
+                    icon: filter.dataType !== logsDataViewType ? 'check' : 'empty',
```
**nit**: this would require an update in case we add new options in the filter such as `metrics` etc, which is something we'd probably want when pushing this down into Discover.
As this entry is used when the `filter.dataType` is undefined, changing this to
```ts
 icon: !filter.dataType ? 'check' : 'empty',
```
should make it work independently from the available types (feel free to invert the condition as you prefer it).

tonyghiani commented on x-pack/plugins/observability_solution/logs_explorer/public/components/data_source_selector/sub_components/data_view_filter.tsx:
```diff
@@ -0,0 +1,109 @@
+              size="xs"
+              onClick={togglePopover}
+            >
+              {filter.dataType !== logsDataViewType ? allDataViewTypesLabel : logsDataViewTypeLabel}
```
**suggestion:** this doesn't scale automatically when we add/remove new filters, we could update it using a map such that:
```ts
function getSelectedFilterLabel(dataType: DataViewsFilterParams['dataType']) {
  const availableFilters = {
    [logsDataViewType]: logsDataViewTypeLabel
  }

  return availableFilters[dataType] || allDataViewTypesLabel
}

// ...

{getSelectedFilterLabel(filter.dataType)}
```

tonyghiani reviewed (approved):
Good job 👏  I left some non-blocking suggestions but feel free to merge it since part of them will be addressed with some parallel work on the selector.

flash1293 commented:
Thanks for the suggestions, fixed the things you brought up and consolidated filter and search into `searchCriteria` in the data views state.

tonyghiani commented on x-pack/plugins/observability_solution/logs_explorer/public/state_machines/data_views/src/defaults.ts:
```diff
@@ -13,7 +13,7 @@ export const createDefaultContext = (): DefaultDataViewsContext => ({
   dataViews: null,
   error: null,
-  search: {
+  searchCriteria: {
```
I might explained it wrong, when I talked about searchCriteria it was not meant to rename the `search` property in the state machine context. My suggestion was about refactoring only the `searchDataViews` function to receive a parameter object as a second argument, but still using it with separate context properties:
```ts
searchDataViews(views, {search: context.search, filter: context.filter})
```
I'd rather keep them separate in the context and keep the `search` naming for consistency with integrations and datasets state machines.

flash1293 replied:
Gotcha, updated
